siriuspy/siriuspy/devices/injsys.py

```diff
@@ -771,3 +773,279 @@ def _command_target(self, cmdtype):
         self._result = retval
 
         self._is_running = ''
+
+
+class InjSysPUModeHandler(_Devices, _Callback):
+    """Device to control pulsed magnets configuration for injection."""
+
+    _DEF_TIMEOUT = 5  # [s]
+    _DEF_SLEEP = 0.1  # [s]
+    SI_DPKCKR_DEFKICK = -6.7  # [mrad]
+    TS_POSANG_DEFDELTA = 2.5  # [mrad]
+    SI_DPKCKR_DLYR_ONAXINC = 28  # [count]
+    SI_DPKCKR_DLYR_OPT = 0  # [count]
+
+    def __init__(self, print_log=True, callback=None):
+        """Init."""
+        self.pudpk = PowerSupplyPU(PowerSupplyPU.DEVICES.SI_INJ_DPKCKR)
+        self.punlk = PowerSupplyPU(PowerSupplyPU.DEVICES.SI_INJ_NLKCKR)
+        self.trigdpk = Trigger('SI-01SA:TI-InjDpKckr')
+        self.trignlk = Trigger('SI-01SA:TI-InjNLKckr')
+        self.posang = PosAng(PosAng.DEVICES.TS)
+        devices = (
+            self.pudpk, self.punlk,
+            self.trigdpk, self.trignlk,
+            self.posang,
+        )
+        self._print_log = print_log
+        self._abort = _Flag()
+
+        # call super init
+        _Devices.__init__(self, '', devices)
+        _Callback.__init__(self, callback=callback)
+
+    @property
+    def is_accum(self):
+        """If configuration is Accumulation."""
+        dpk_off = (self.pudpk.pwrstate == PowerSupplyPU.PWRSTATE.Off) or \
+            (self.pudpk.pulse == PowerSupplyPU.PULSTATE.Off)
+        nlk_on = (self.punlk.pwrstate == PowerSupplyPU.PWRSTATE.On) and \
+            (self.punlk.pulse == PowerSupplyPU.PULSTATE.On)
+        return dpk_off & nlk_on
+
+    def cmd_switch_to_accum(self):
+        """Switch to Accumulation."""
+        self._abort.clear()
+        self._update_status('Switching to PU Accumulation config...')
+        if not self.pudpk.interlock_ok:
+            self._update_status('ERR: DpK interlocks not ok. Aborted.')
+            return False
+        if not self.punlk.interlock_ok:
+            self._update_status('ERR: NLK interlocks not ok. Aborted.')
+            return False
+        if self._abort.is_set():
+            self._clr_flag_abort()
+            return False
+
+        # if previously in on-axis, do delta angle x
+        if self.is_onaxis:
```

Comment:
The block after this if statement also seems to be repeated along the code, except for the value of delta applied in posang.delta_angx. Maybe creating a method to do this process that takes as input the angle delta to be applied in posang might be good

Reply:
Done!